lib/class-google.php

```diff
@@ -0,0 +1,321 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+/**
+ * Google Webfonts API class.
+ *
+ * Contains integration with Google Webfonts Developer API.
+ */
+class Google extends \DPP\Admin\Fonts\Source {
+
+	/**
+	 * Google API key.
+	 *
+	 * @var string
+	 */
+	protected $api_key = '';
+
+	/**
+	 * Fon source name.
+	 *
+	 * @var string
+	 */
+	protected $name = 'google';
+
+	/**
+	 * Handle our checks then call our hooks.
+	 *
+	 * @return void
+	 */
+	public function init() {
+
+		// First make sure we have our main class. not sure how we wouldn't but then again...
+		if ( ! class_exists( 'Genesis_Palette_Pro' ) ) {
+			return;
+		}
+
+		// Get the API key from the options table if it exists.
+		$this->api_key = $this->get_api_key();
+
+		parent::init();
+
+		add_action( 'admin_notices', array( $this, 'api_key_action_response' ) );
+
+		// Enqueue any scripts that the source may need.
+		add_action( 'wp_enqueue_scripts', array( $this, 'font_scripts' ) );
+
+		add_action( 'dpp_preview_head_before_styles', array( $this, 'preview_head' ) );
+	}
+
+	/**
+	 * Set the font source config.
+	 *
+	 * @param array $defaults Default config parameters.
+	 *
+	 * @return void
+	 */
+	protected function set_config( $defaults ) {
+		$this->config = array(
+			'label' => __( 'Google Webfonts', 'gppro-google-webfonts' ),
+		);
+
+		$this->config = wp_parse_args( $this->config, $defaults );
+	}
+
+	/**
+	 * Get the API Key from the options table.
+	 *
+	 * @return string
+	 */
+	protected function get_api_key() {
+		if ( empty( $this->api_key ) ) {
+			// Get the API key from the options table if it exists.
+			$this->api_key = get_option( 'gppro_google_webfonts_api_key', '' );
+		}
+
+		return $this->api_key;
+	}
+
+	/**
+	 * Check to make sure the api key is valid.
+	 *
+	 * @param string $key The api key.
+	 *
+	 * @return boolean
+	 */
+	protected function api_key_check( $key ) {
+		if ( '' !== $key ) {
+			$fonts = $this->get_fonts( array() );
+
+			if ( false !== $fonts ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Load the list of fonts available from the soruce.
+	 *
+	 * @return void
+	 */
+	protected function load_fonts() {
+		if ( empty( $this->fonts ) ) {
+			$key = $this->api_key;
+
+			if ( '' === $key ) {
+				return false;
+			}
+
+			$response = wp_remote_get( esc_url( 'https://www.googleapis.com/webfonts/v1/webfonts?key=' . $key ) );
+
+			if ( is_array( $response ) ) {
+				if ( isset( $response['response']['code'] ) && 200 === $response['response']['code'] ) {
+					$response_fonts = json_decode( $response['body'] )->items;
+					$fonts          = array();
+
+					array_walk(
+						$response_fonts,
+						function( $font, $font_index ) use ( &$fonts ) {
```

Comment:
can/should we make this a declared function instead of an anonymous one? This is getting pretty deep (indentation wise) so I wonder if that will help improve readability and potentially testability.

Reply:
Done.